src/charm.py

```diff
@@ -3,258 +3,124 @@
-            return True, ""
-
-        return (
-            False,
-            "Missing required configuration: "
-            f"{' '.join(sorted(required_options - non_empty_options))}",
-        )
-
-    def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
-        """Set relation data for the unit once an agent has connected.
-
-        Args:
-            event: information about the relation joined event.
-        """
-        logger.info("Jenkins relation joined")
-        num_executors = os.cpu_count()
-        config_labels = self.model.config.get("jenkins_agent_labels")
-
-        if config_labels:
-            labels = config_labels
-        else:
-            labels = os.uname().machine
-
-        event.relation.data[self.model.unit]["executors"] = str(num_executors)
-        event.relation.data[self.model.unit]["labels"] = labels
-        event.relation.data[self.model.unit]["slavehost"] = self._stored.relation_agent_name
-
-    def _on_agent_relation_changed(self, event: RelationChangedEvent) -> None:
-        """Populate local configuration with data from relation.
-
-        Args:
-            event: information about the relation changed event.
-        """
-        logger.info("Jenkins relation changed")
-
-        # Check event data
-        try:
-            # Right side data type is Union[Unit, Application]
-            relation_jenkins_url = event.relation.data[event.unit]["url"]  # type: ignore
-        except KeyError:
-            logger.warning(
-                "Expected 'url' key for %s unit in relation data. Skipping setup for now.",
-                event.unit,
+        if not self.state.jenkins_config:
+            if self.model.get_relation(SLAVE_RELATION):
+                self.model.unit.status = ops.BlockedStatus(
+                    "Please remove and re-relate slave relation."
+                )
+                return
+            # Support fallback relation to AGENT_RELATION.
+            self.model.unit.status = ops.BlockedStatus(
+                "Please remove and re-relate agent relation."
             )
-            self.model.unit.status = ActiveStatus()
             return
+
         try:
-            # Right side data type is Union[Unit, Application]
-            relation_secret = event.relation.data[event.unit]["secret"]  # type: ignore
-        except KeyError:
-            logger.warning(
-                "Expected 'secret' key for %s unit in relation data. Skipping setup for now.",
-                event.unit,
+            server.download_jenkins_agent(
+                server_url=self.state.jenkins_config.server_url,
+                container=container,
             )
-            self.model.unit.status = ActiveStatus()
+        except server.AgentJarDownloadError as exc:
+            logger.error("Failed to download Agent JAR executable, %s", exc)
+            self.model.unit.status = ops.ErrorStatus("Failed to download Agent JAR executable.")
```

Comment:
Yup, it is intended to crash and error since this isn't quite a recoverable (if it is recoverable via user action it should be blocked status)

Reply:
In that case, instead of displaying the error message `Failed to download Agent JAR executable.` in the Juju status, a more general error, `hook failed: "config-changed"` would appear. Would this be considered acceptable?

Reply:
Right, I think that's a good point!
However, I've received [this feedback](https://github.com/canonical/jenkins-agent-k8s-operator/pull/26#discussion_r1251707017)
> Is there anything the user can do to unblock ? If so, we should suggest in the message what to do, if not, we should reraise the exception

It would be better to direct users to a better solution but that's a bit of a black box right now. I think we could update the error message when we are aware of the behaviors that may make the Jar donwload fail. (network, server unavaialble, ...etc perhaps?)

Reply:
If your preference is for the charm to transition into an error state, I recommend directly raising an exception such as `RuntimeError` within the charm itself. Setting the status to ErrorStatus may appear to change the status to an error state, while in reality it's not possible.

Reply:
Changed to re-raise from exception which gives more visibility in logging, thank you for the explanation!